common/persistence/queue_v2_util.go

```diff
@@ -54,22 +56,92 @@ func GetNextPageTokenForQueueV2(result []QueueV2Message) []byte {
 	return append([]byte{pageTokenPrefixByte}, b...)
 }
 
-func GetMinMessageIDForQueueV2(request *InternalReadMessagesRequest) (int64, error) {
-	// TODO: start from the ack level of the queue partition instead of the first message ID when there is no token.
-	if len(request.NextPageToken) == 0 {
-		return FirstQueueMessageID, nil
+func GetMinMessageIDToReadForQueueV2(
+	queueType QueueV2Type,
+	queueName string,
+	nextPageToken []byte,
+	queue *persistencespb.Queue,
+) (int64, error) {
+	if len(nextPageToken) == 0 {
+		partition, err := GetPartitionForQueueV2(queueType, queueName, queue)
+		if err != nil {
+			return 0, err
+		}
+		return partition.MinMessageId, nil
 	}
 	var token persistencespb.ReadQueueMessagesNextPageToken
 
 	// Skip the first byte. See the comment on pageTokenPrefixByte for more details.
-	err := token.Unmarshal(request.NextPageToken[1:])
+	err := token.Unmarshal(nextPageToken[1:])
 	if err != nil {
 		return 0, fmt.Errorf(
 			"%w: %q: %v",
 			ErrInvalidReadQueueMessagesNextPageToken,
-			request.NextPageToken,
+			nextPageToken,
 			err,
 		)
 	}
 	return token.LastReadMessageId + 1, nil
 }
+
+func GetPartitionForQueueV2(
+	queueType QueueV2Type,
+	queueName string,
+	queue *persistencespb.Queue,
+) (*persistencespb.QueuePartition, error) {
+	// Currently, we only have one partition for each queue. However, that might change in the future. If a queue is
+	// created with more than 1 partition by a server on a future release, and then that server is downgraded, we
+	// will need to handle this case. Since all DLQ tasks are retried infinitely, we just return an error.
+	numPartitions := len(queue.Partitions)
+	if numPartitions != 1 {
+		return nil, serviceerror.NewInternal(
+			fmt.Sprintf(
+				"queue with type %v and queueName %v has %d partitions, but this implementation only supports"+
+					" queues with 1 partition. Did you downgrade your Temporal server?",
+				queueType,
+				queueName,
+				numPartitions,
+			),
+		)
+	}
+	partition := queue.Partitions[0]
+	return partition, nil
+}
+
+type DeleteRequest struct {
+	// LastIDToDeleteInclusive represents the maximum message ID that the user wants to delete, inclusive.
+	LastIDToDeleteInclusive int64
+	// ExistingMessageRange represents an inclusive range of the minimum message ID and the maximum message ID in the queue.
+	ExistingMessageRange InclusiveMessageRange
+}
+
+type InclusiveMessageRange struct {
+	MinMessageID int64
+	MaxMessageID int64
+}
+
+type DeleteRange struct {
+	InclusiveMessageRange
+	NewMinMessageID int64
+}
+
+// GetDeleteRange returns the range of messages to delete, and a boolean indicating whether any messages should be deleted.
+func GetDeleteRange(request DeleteRequest) (DeleteRange, bool) {
+	lastID := request.LastIDToDeleteInclusive
+	newMinMessageID := lastID + 1
+	if lastID >= request.ExistingMessageRange.MaxMessageID {
+		// Never actually delete the last message
+		lastID = request.ExistingMessageRange.MaxMessageID - 1
+	}
```

Comment:
I think there's actually still a bug here. For example, if the messages are 0,1,2, and the requested last ID is 4, then we'll delete messages 0,1, and set min_message_id to 5. Then, message 2 will become a zombie message that can never be deleted, and the next message will have id 5, which is too high. Can you fix and also add a test for this case?

Reply:
Good catch! The actual issue is when we read messages. After the deletion (with request last ID 4), if we enqueue a message, it will still have ID 3. Then if we try to read messages, we will never get this message. (Because min message id is 5).  So we will lose messages in the queue.  Fixed the issue. I have refactored GetDeleteRange a little bit to make it bit more concise.